www/blog/2023-01-08-tinyrpc-client.mdx

```diff
@@ -0,0 +1,291 @@
+
+```twoslash include decoraterecord
+import type { AnyRouter, ProcedureRouterRecord } from "@trpc/server";
+
+type DecoratedProcedureRecord<TProcedures extends ProcedureRouterRecord> = {
+  [TKey in keyof TProcedures]: TProcedures[TKey] extends AnyRouter
+    ? DecoratedProcedureRecord<TProcedures[TKey]['_def']['record']>
+    : TProcedures[TKey] extends AnyProcedure
+    ? DecorateProcedure<TProcedures[TKey]>
+    : never;
+};
+```
+
+```ts twoslash
+// @include: resolver
+// @include: decorate
+// ---cut---
+// @include: decoraterecord
+```
+
+Let's digest this type a bit:
+
+1. We pass a `ProcedureRouterRecord` to the type as a generic, which is a type containing all the procedures and sub-routers that exists on a tRPC router.
+2. We iterate over the keys of the record, which are the procedure or router names, and do the following:
+   - If the key maps to a router, we recursively call the type on that router's procedure record, which will decorate all the procedures in that router. This will provide autocompletion as we traverse the path.
+   - If the key maps to a procedure, we decorate the procedure using the `DecorateProcedure` type we created earlier.
+   - If the key doesn't map to a procedure or router, we assign the `never` type which is like saying "this key doesn't exist" which will cause a type error if we try to access it.
+
+Finally, we'll need to use these types as we create the client. We'll add a generic to the `createTinyRPCClient` function that will accept any tRPC router type, and then we'll cast the return type of the function to the `DecoratedProcedureRecord` type we created earlier:
+
+```ts twoslash
+// @include: proxy
+// @include: resolver
+// @include: decorate
+// @include: decoraterecord
+// ---cut---
+export const createTinyRPCClient = <TRouter extends AnyRouter>(url: string) => {
+  return createRecursiveProxy(async (opts) => {
+    // ...
+  }, []) as DecoratedProcedureRecord<TRouter['_def']['record']>;
+};
+```
+
+And that's it! Create the client and provide it your server's url and you'll get full autocompletion and type safety when you call your procedures!
+
+```ts twoslash
+// @target: esnext
+// @filename: tinyrpc.ts
+// @include: proxy
+// @include: resolver
+// @include: decorate
+// @include: decoraterecord
+export const createTinyRPCClient = <TRouter extends AnyRouter>(url: string) => {
+  return createRecursiveProxy(async (opts) => {
+    // ...
+  }, []) as DecoratedProcedureRecord<TRouter['_def']['record']>;
+};
+
+// @errors: 2339
+
+// @filename: router.ts
+// @include: router
+// @filename: client.ts
+// ---cut---
+import type { AppRouter } from './router';
+import { createTinyRPCClient } from './tinyrpc';
+
+const client = createTinyRPCClient<AppRouter>('http://localhost:3000/api/trpc');
+//    ^?
+
+// 🧙‍♀️ magic autocompletion
+client.post.b;
+//           ^|
+
+// 👀 fully typesafe
+const post = await client.post.byId.query({ id: '123' });
+//     ^?
```

Comment:
![CleanShot 2023-01-12 at 23 48 09](https://user-images.githubusercontent.com/51714798/212197751-16689b93-7fd8-40f7-8ad0-4d362c52c77e.png)

no clue why this isn't two slashing :(